src/main/scala/scalafix/sbt/ScalafixPlugin.scala

```diff
@@ -88,6 +96,29 @@ object ScalafixPlugin extends AutoPlugin {
     def scalafixConfigSettings(config: Configuration): Seq[Def.Setting[_]] =
       Seq(
         scalafix := scalafixInputTask(config).evaluated,
+        compile := {
+          val oldCompile = compile.value
+          val runScalafix = scalafixOnCompile.value
+          val currentState = state.value
+          val ref = thisProjectRef.value
+          if (runScalafix) {
+            // We can't just run scalafix from here, as it would introduce a cyclic dependency
+            // since semantic rules require compilation products. Instead we run scalafix within
+            // a custom state where compile has not been overriden.
+            val extracted = Project.extract(currentState)
+            val withOldCompile = Compat.append(
+              extracted,
+              Seq(compile.in(ref, config) := oldCompile),
+              currentState
+            )
+            extracted.runInputTask(
+              scalafix.in(ref, config),
+              input = "",
+              withOldCompile
+            )
```

Comment:
This works, but it is very hacky, as the state "reload" is slow, and compilation runs for a given project/config are not de-duped during the liftetime of task execution... Still trying to figure out something better.

Reply:
> 2\. compilations for a given project/config and most annoyingly all the projects it depends on are not de-duped during the liftetime of a task execution

Yes, because `runInputTask` is running outside of the task graph, which is dangerous for race conditions.  See these references for when sbt-buildinfo used to do the same thing, and the problems it caused: https://github.com/sbt/sbt-buildinfo/commit/7f3a548075b71b261cddc60fd6143ff7dfb87bf7#diff-0a369498a5a8db3ac8fa606b544c9810

Reply:
Thanks for the pointers, that confirms my concerns! How did you get rid of it there? I am just wondering if it would be applicable here.

Reply:
I changed the implementation so the values were defined within the Task (monad): https://github.com/sbt/sbt-buildinfo/compare/v0.7.0...v0.9.0#diff-60231996cd2e40cc694705deb78dc6d6R27; notice the `case BuildInfoKey.TaskValue`, which doesn't do the unsafe `extracted.runTask` (the `case BuildInfoKey.Task` is still there for backwards-compatibility and extreme cases.)